internal/controller/serving/inferenceservice_controller.go

```diff
@@ -132,8 +137,41 @@ func (r *InferenceServiceReconciler) ReconcileServing(ctx context.Context, req c
 		return ctrl.Result{}, err
 	}
 
-	if isvc.GetDeletionTimestamp() != nil {
-		return reconcile.Result{}, r.onDeletion(ctx, logger, isvc)
+	if isvc.GetDeletionTimestamp() == nil {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		logger.Info("Adding Finalizer")
+		if !controllerutil.ContainsFinalizer(isvc, constants.InferenceServiceODHFinalizerName) {
+			controllerutil.AddFinalizer(isvc, constants.InferenceServiceODHFinalizerName)
+			patchYaml := "metadata:\n  finalizers: [" + strings.Join(isvc.ObjectMeta.Finalizers, ",") + "]"
+			patchJson, _ := yaml.YAMLToJSON([]byte(patchYaml))
+			if err := r.Patch(ctx, isvc, client.RawPatch(types.MergePatchType, patchJson)); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		var deleteErrors *multierror.Error
+		logger.Info("InferenceService being deleted")
+		if controllerutil.ContainsFinalizer(isvc, constants.InferenceServiceODHFinalizerName) {
+			err := r.onDeletion(ctx, logger, isvc)
+			if err != nil {
+				deleteErrors = multierror.Append(deleteErrors, err)
+			}
+			// Check if we need to also perform cleanup on the namespace
+			err1 := r.DeleteResourcesIfNoIsvcExists(ctx, logger, req.Namespace)
```

Comment:
Now that this is being invoked here, should the call in line 130 be removed?

Reply:
I kept it to catch the scenario where a user might manually remove the finalizer by editing the isvc object before odh-model-controller has had a chance to process everything (but this is a very unlikely scenario). What are your thoughts?

Reply:
I, personally, prefer to remove. If they manually remove the finalizer, users must know what they are doing.

Reply:
So, it is your call. It doesn't hurt.

Reply:
I think let's keep it